dotnet/Gherkin/GherkinLine.cs

```diff
@@ -1,169 +1,316 @@
+    public string GetLineText(int indentToRemove = 0)
     {
         if (indentToRemove < 0 || indentToRemove > Indent)
-            return trimmedLineText;
+            return lineText.Substring(trimmedStartIndex);
 
         return lineText.Substring(indentToRemove);
     }
 
+    /// <summary>
+    /// Returns the remaining part of the line.
+    /// </summary>
+    /// <param name="length">Number of characters to skip from the beginning of the line, including indentation.</param>
+    /// <returns></returns>
     public string GetRestTrimmed(int length)
     {
-        return trimmedLineText.Substring(length).Trim();
+        return lineText.Substring(trimmedStartIndex + length).Trim();
     }
 
-    public IEnumerable<GherkinLineSpan> GetTags()
+    public readonly struct TagsEnumerable : IEnumerable<GherkinLineSpan>
     {
-        string uncommentedLine = trimmedLineText;
-        var commentIndex = trimmedLineText.IndexOf(GherkinLanguageConstants.COMMENT_PREFIX[0]);
-        while (commentIndex >= 0)
+        readonly int lineNumber;
+        readonly string uncommentedLine;
+        readonly int position;
+        public TagsEnumerable(int lineNumber, string lineText, int trimmedStartIndex)
         {
-            if (commentIndex == 0)
-                yield break;
-            if (Array.IndexOf(inlineWhitespaceChars, trimmedLineText[commentIndex - 1]) == 0)
+            this.lineNumber = lineNumber;
+            uncommentedLine = lineText;
+            var commentIndex = lineText.IndexOf(GherkinLanguageConstants.COMMENT_PREFIX[0], trimmedStartIndex);
+            while (commentIndex >= 0)
             {
-                uncommentedLine = uncommentedLine.Substring(0, commentIndex);
-                break;
+                if (commentIndex == 0)
+                {
+                    position = -1;
+                    return;
+                }
+                if (Array.IndexOf(inlineWhitespaceChars, lineText[commentIndex - 1]) != -1)
+                {
+                    uncommentedLine = uncommentedLine.Substring(0, commentIndex);
+                    break;
+                }
+                commentIndex = lineText.IndexOf(GherkinLanguageConstants.COMMENT_PREFIX[0], commentIndex + 1);
             }
-            commentIndex = trimmedLineText.IndexOf(GherkinLanguageConstants.COMMENT_PREFIX[0], commentIndex + 1);
+            position = uncommentedLine.IndexOf(GherkinLanguageConstants.TAG_PREFIX[0], trimmedStartIndex);
         }
-        int position = Indent;
-        foreach (string item in uncommentedLine.Split(GherkinLanguageConstants.TAG_PREFIX[0]))
+
+        public TagsEnumerator GetEnumerator() => new TagsEnumerator(lineNumber, uncommentedLine, position);
+
+        IEnumerator<GherkinLineSpan> IEnumerable<GherkinLineSpan>.GetEnumerator() => GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+    }
+
+    public struct TagsEnumerator : IEnumerator<GherkinLineSpan>

```

Comment:
Like above, do we need to make this public? (Also `TableCellsEnumerator`)

Reply:
See above.